src/Servers.jl

```diff
@@ -410,104 +307,79 @@ function handle_connection(f::Function, c::Connection;
+            finally
+                close(c)
+            end
+            break
+        end
+        sleep(8 + rand() * 4)
+    end
+    return
+end
 
 """
-Create a `HTTP.Stream` and parse the Request headers from a `HTTP.Transaction`.
-If there is a parse error, send an error Response.
-Otherwise, execute stream processing function `f`.
+Transaction handler: creates a new Stream for the Transaction, calls startread on it,
+then dispatches the stream to the user-provided handler function. Catches errors on all
+IO operations and closes gracefully if encountered.
 """
-function handle_transaction(f::Function, t::Transaction;
-                            final_transaction::Bool=false,
-                            verbose::Bool=false, kw...)
-
-    request = HTTP.Request()
-    http = Streams.Stream(request, t)
+function handle(h, t::Transaction, last::Bool=false)
+    request = Request()
+    http = Stream(request, t)
 
     try
         startread(http)
     catch e
-        # @show typeof(e)
-        # @show fieldnames(e)
         if e isa EOFError && isempty(request.method)
             return
-# FIXME https://github.com/JuliaWeb/HTTP.jl/pull/178#pullrequestreview-92547066
-#        elseif !isopen(http)
-#            @warn "Connection closed"
-#            return
-        elseif e isa HTTP.ParseError
-            @error e
+        elseif e isa ParseError
             status = e.code == :HEADER_SIZE_EXCEEDS_LIMIT  ? 413 : 400
             write(t, Response(status, body = string(e.code)))
             close(t)
             return
+        elseif e isa Base.IOError && e.code == -54
+            # read: connection reset by peer (ECONNRESET)
+            return
         else
             rethrow(e)
         end
     end
 
-    if verbose
-        @info http.message
-    end
-
-    response = request.response
-    response.status = 200
-    if final_transaction || hasheader(request, "Connection", "close")
-        setheader(response, "Connection" => "close")
+    request.response.status = 200
+    if last || hasheader(request, "Connection", "close")
+        setheader(request.response, "Connection" => "close")
     end
 
     @async try
-        handle_stream(f, http)
-    catch e
-        if isioerror(e)
-            @warn e
-        else
-            @error exception=(e, stacktrace(catch_backtrace()))
-        end
-        close(t)
-    finally
-        final_transaction && close(t.c)
-    end
-    return
-end
-
-"""
-Execute stream processing function `f`.
-If there is an error and the stream is still open,
-send a 500 response with the error message.
-
-Close the `Stream` for read and write (in case `f` has not already done so).
-"""
-function handle_stream(f::Function, http::Stream)
-
-    try
-        if applicable(f, http)
-            f(http)
-        else
-            handle_request(f, http)
-        end
+        handle(h, http)
+        closeread(http)
+        closewrite(http)
```

Comment:
Is it ok for these close functions to be skipped when handle throws an error?
I guess it is because the catch sets `last` and the finally closes the underlying connection on `last` at which point the read and write state is moot.